File: src/weaponinventorymod/gui/WimGuiSemanticButtonFactory.java
package weaponinventorymod.gui;

import java.awt.Color;

final class WimGuiSemanticButtonFactory<A> {
    private final Color borderColor;

    WimGuiSemanticButtonFactory(Color borderColor) {
        this.borderColor = borderColor;
    }

    WimGuiButtonSpec<A> button(float width,
                               String label,
                               A action,
                               boolean enabled,
                               Color fillColor) {
        return button(width, label, action, enabled, fillColor, null);
    }

    WimGuiButtonSpec<A> button(float width,
                               String label,
                               A action,
                               boolean enabled,
                               Color fillColor,
                               String tooltip) {
        return WimGuiButtonSpec.semantic(width, label, action, enabled, fillColor, borderColor, tooltip);
    }

    WimGuiButtonSpec<A> enabledButton(float width,
                                      String label,
                                      A action,
                                      Color fillColor) {
        return button(width, label, action, true, fillColor);
    }

    WimGuiButtonSpec<A> enabledButton(float width,
                                      String label,
                                      A action,
                                      Color fillColor,
                                      String tooltip) {
        return button(width, label, action, true, fillColor, tooltip);
    }
}

File: src/main/kotlin/weaponsprocurement/ui/WimGuiSemanticButtonFactory.kt
package weaponsprocurement.ui

import java.awt.Color

class WimGuiSemanticButtonFactory<A>(private val borderColor: Color) {
    fun button(width: Float, label: String?, action: A, enabled: Boolean, fillColor: Color): WimGuiButtonSpec<A> =
        button(width, label, action, enabled, fillColor, null)

    fun button(
        width: Float,
        label: String?,
        action: A,
        enabled: Boolean,
        fillColor: Color,
        tooltip: String?,
    ): WimGuiButtonSpec<A> = WimGuiButtonSpec.semantic(width, label, action, enabled, fillColor, borderColor, tooltip)

    fun enabledButton(width: Float, label: String?, action: A, fillColor: Color): WimGuiButtonSpec<A> =
        button(width, label, action, true, fillColor)

    fun enabledButton(
        width: Float,
        label: String?,
        action: A,
        fillColor: Color,
        tooltip: String?,
    ): WimGuiButtonSpec<A> = button(width, label, action, true, fillColor, tooltip)
}
